### DISCUSSION TITLE
Add better window snapping/management

### DISCUSSION THREAD

**Discussion by NotoriousPyro** [OpenRCT2/OpenRCT2#16993] [Feature requests]
With a super ultrawide screen, the dialog windows that open when clicking a ride or a visitor seemingly go to whatever side of the screen they want to, thus making it a bit hard to play with such a wide screen on the neck without having to set the game to Windowed.

I think a better way would be to have the windows open in a predetermined set of locations, for example, if the screen is as big as an ultrawide it could use only 50-60% of the screen in the center, and start stacking in the top left of that area. Alternatively, the windows could open by default in the center, and then "push" other windows to the side as new ones are opened. At the moment, the game is quite difficult to play since I don't know where a window will open and it is tedious to click each little 'x' all over the screen.

**Comment by ocalhoun6** [comment]
Maybe an acceptable solution -- and something that would be beneficial in a lot of other use cases besides -- would be to implement some kind of window position memory?

If the game remembered the last place you put each type of window, then defaulted it to placing it in the same position next time (or as near as possible if other windows are in the way), that would mostly solve this issue. Once you'd positioned your windows to your liking on your ultra-wide screen, you wouldn't have to worry about windows opening in weird places afterward.

It would also be a nice customization/quality of life improvement for anybody who doesn't like the default window placements.

(Maybe also have a 'remember window position' flag that gets set only when a window is moved manually. And the window position memory will only record the new position if a window is closed while that flag is set. This would prevent automatic window movements -- such as when moving windows to make room for other windows -- from being recorded as the new default, so that automatic window movements wouldn't mess up someone's carefully customized window placements.)

Though, for the sake of simplicity, this window position memory system would probably ignore any windows created by plugins? Because there might be a huge variety of different windows created by plugins, which would make keeping track of all those a huge headache.

And, of course, you'd have to have some code in the system that overrides the remembered window position back to the default in case the resolution has changed and the window would no longer be visible on screen if it was put in the remembered position.

**Comment by NotoriousPyro** [comment]
The problem with that suggestion is it means I still have to arrange things, but this still has the original problem... The windows open starting position appears random, it seems it would be better if it was more predictable.

An alternative solution would be to have tabbing, if you open more than one of the same type of tab, then combine them into one window (but with a tab description of what each one contained, e.g. the ride name). That way I don't need to ever close any of them.

**Reply by ocalhoun6** [reply]
>That way I don't need to ever close any of them.

Well, that might work with a huge ultra-wide ... but normal-sized screens would be completely covered with different types of windows if you tried this approach.

**Reply by NotoriousPyro** [reply]
Not if they're stacked in a tabbed view.

**Reply by NotoriousPyro** [reply]
The main problem is the window appearance is always random. If I close that 'ride' window and reopen it, it will open in a corner but I'm looking at the center of the screen 99% of the time, most as anyone. I don't mind particularly the buttons at the top are either side of the screen, but it would be nice to have them pushed more centrally for bigger screens.

However, the main problem is that the windows open in unpredictable places meaning my neck can get sore placing this game at those resolutions (however great to see so much park), does not work really well.

**Reply by NotoriousPyro** [reply]
The only workaround right now I have is to have OpenRCT2 in a Window that uses about 60% of the screen. I suppose it works, but I want to be able to run it fullscreen for immersion :).

I know very little about C++ otherwise I would try to raise a PR for it myself with the idea.

**Reply by ocalhoun6** [reply]
>The main problem is the window appearance is always random.

Well, if you had window position memory, it wouldn't be random. It would just be in the last place you put it.

You'd have to arrange the windows to your liking _once_, but after that, you'd be golden.

**Reply by NotoriousPyro** [reply]
Still, that isn't what I think is going to solve the problem. I have to do it for each new game then right? Since each ride would be a new entity? Whereas, a system which would place them centrally at the top on-screen and then push them side-by-side as more appear would work regardless of if the window has been seen or not. I don't really care for it to remember the position of them, I just want a predictable place to see it and to be able to close it with similar predictability.

**Reply by NotoriousPyro** [reply]
I suppose what you're suggesting is somewhat alleviating to the problem though.

**Reply by ocalhoun6** [reply]
>I have to do it for each new game then right? Since each ride would be a new entity?

No, "ride window" would be one entity, and every time you open any ride window in any game, it would try to put that ride window in the same place as the previous ride window you had open, even if that window was in a previous game. You'd literally only have to move the windows where you want them _once_ -- once _ever_.
